### ISSUE TITLE
Copy Link to Paragraph / to Page throw an error

### ISSUE THREAD

**Issue by jasonjac2** [stevencohn/OneMore#995]
# Problem to Solve

Installed 5.8.3 and I get
![image](https://user-images.githubusercontent.com/68282090/226382334-2f76dbc7-777d-485d-aa31-5997c070b4ef.png)

ON's own copy link to para works fine as does normal copy.

I haven't rebooted yet, but will do. 
08| Log restarted
05| Running command CopyLinkCommand
24| Error running command CopyLinkCommand
24| System.Runtime.InteropServices.COMException
Message: OpenClipboard Failed (Exception from HRESULT: 0x800401D0 (CLIPBRD_E_CANT_OPEN))
Description: Unrecognized error code 0x800401D0
ErrorCode: 0x800401D0 (-2147221040)
HResult: 0x800401D0 (-2147221040)
Source: mscorlib
StackTrace:    at System.Runtime.InteropServices.Marshal.ThrowExceptionForHRInternal(Int32 errorCode, IntPtr errorInfo)
   at System.Windows.Clipboard.Flush()
   at System.Windows.Clipboard.CriticalSetDataObject(Object data, Boolean copy)
   at System.Windows.Clipboard.SetDataObject(Object data, Boolean copy)
   at River.OneMoreAddIn.ClipboardProvider.<RestoreState>b__10_0()
   at River.OneMoreAddIn.SingleThreaded.<>c__DisplayClass1_0.<Invoke>b__0()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
   at River.OneMoreAddIn.ClipboardProvider.<RestoreState>d__10.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
   at River.OneMoreAddIn.CopyLinkCommand.<Execute>d__2.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
   at River.OneMoreAddIn.CommandFactory.<Run>d__5.MoveNext()
TargetSite: [mscorlib] System.Runtime.InteropServices.Marshal::ThrowExceptionForHRInternal()

# To Reproduce

# Expected Behavior

# Screenshots

# Environment (if applicable)

 - OneNote Version: [e.g. Office 2019, O365, also 32bit or 64bit?]
 - OneMore Version [e.g. 3.4 - from the About dialog or from the log file]

# Additional Context

Add any other context about the problem here. Attach page XML file or the app log file as appropriate. You can find the log file here: %localappdata%\temp\OneMore.log

THE FINE PRINT This is a hobby project that I started for myself. I'm happy to take requests and will promise to
consider each one carefully but offer no guarantees that I will ultimately agree to their
usefulness or have time to implement any of them. If OneNote offers a "close enough" work-around
then I will probably reject the request - you've been warned!

**Comment by stevencohn** [comment]
I was digging a little more since I wasn't able to repro the problem. I found a couple of articles that talk about locking the clipboard when copy/pasting across RDP sessions, or that screen clipping and recording software might incorrectly lock the clipboard. 

Are you running anything like that? I can add a retry. But I also can't seem to repro it to test whether that would fix it.

**Comment by jasonjac2** [comment]
The plot thickens.  So - reboot.  I think It worked once, but now when I use the OneMore, copy page, copy para nothing happens, no warning, to confirmation, no clipboard action.  i.e. clipboard stays the same.

OneMore navigator also looks frozen (I keep it up in the background) [I have set History to 100 pages]

**Comment by stevencohn** [comment]
A couple of folks mentioned the slowness in #992. Don't worry about a ticket, it seems real so I'll take a look asap, both at the slowness and the clipboard. Thanks

**Comment by stevencohn** [comment]
The slowness has nothing to do with the Navigator window. I'll release a hotfix later today.

**Comment by jasonjac2** [comment]
I've narrowed this down to a tool I have written that monitors the clipboard.  If it is active, then Version 5.8.4 with OneNote 16.0.16130.20306 throws the error. 

The app waits till the clipboard changes, then grabs a copy of the .AsText and then processes that completely independently of the clipboard (basically slices and dices the text). I have added extra debugging to ensure my method enter and exit appropriately without failing, which they do.  Also they don't take a lot of time. 
They do this though... brace yourself for some pascal:
``` Pascal 
function TClipboard.GetAsText: string;
var
  Data: THandle;
begin
  Open;
  Data := GetClipboardData(CF_TEXT);
  try
    if Data <> 0 then
      Result := PChar(GlobalLock(Data))
    else
      Result := '';
  finally
    if Data <> 0 then GlobalUnlock(Data);
    Close;
  end;
end;
```

Open and close appear to be reference counting:
``` Pascal 
procedure TClipboard.Open;
begin
  if FOpenRefCount = 0 then
  begin
    FClipboardWindow := Application.Handle;
    if FClipboardWindow = 0 then
    begin
      FClipboardWindow := Classes.AllocateHWnd(MainWndProc);
      FAllocated := True;
    end;
    if not OpenClipboard(FClipboardWindow) then
      raise Exception.CreateRes(@SCannotOpenClipboard);
    FEmptied := False;
  end;
  Inc(FOpenRefCount);
end;
```

GlobalLock appears to be a Win32 Kernel32.dll call.

**Comment by jasonjac2** [comment]
I wonder why none of the native copies from OneNote have the same issue.  I would be more than happy to send over my little app.  With no data file it doesn't do a lot, but there is a tab for monitoring the clipboard and then parsing the text looking for [*] and adding the content to buttons.  It doesn't do any "funny business".  Basically it allows me to quickly filter csv file.  I have added a "OutputDebugStrings"  for poor man debugging as I was initially concerned that my app was throwing an exception and somehow breaking the  message event model expected by clipboard interaction.

**Comment by stevencohn** [comment]
I haven't used Pascal in many decades but used to love it! But I wouldn't be able to compile and debug it.

Anyway, it's unclear to me, just for the code snippets, what the relationship is between your GetAsText and Open methods. You're not sharing FClipboardWindow.

How does your app hook into and watch clipboard? Is it polling on a regular interval, is it through some eventing mechanism, or is it triggered from a button? Do you ever call GlobalUnlock?

**Comment by jasonjac2** [comment]
No polling.  I would imagine that windows broadcasts a message when the clipboard changes OR you register an interest and then the first "window" get the message and passes it on through the windows messaging system.

I'll have to come back to it unfortunately as here (coming up to April 5th) all developers  of financial systems are under the cosh! :-).

I'll look into it further later.  Sorry to stop the flow.

For the moment all is good as long as I turn off the clipboard monitoring. 

Before I go:
GlobalUnlock is in the finally, i.e. assuming data is still <> 0 it would execute.

``` pascal
procedure TJvClipboardMonitor.SetEnabled(Value: Boolean);
begin
  if FEnabled <> Value then
  begin
    if Value then
      FNextWindow := SetClipboardViewer(FWindowHandle) // windows API.
    else
    begin
      ChangeClipboardChain(FWindowHandle, FNextWindow);
      FNextWindow := 0;
    end;
    FEnabled := Value;
  end;
end;

procedure TJvClipboardMonitor.WndProc(var AMsg: TMessage);
begin
  with AMsg do
  begin
    Result := 0;
    case Msg of
      WM_DESTROYCLIPBOARD:
        ClipboardChanged;
      WM_CHANGECBCHAIN:
        if HWND(WParam) = FNextWindow then
          FNextWindow := HWND(LParam)
        else
          ForwardMessage(AMsg);
      WM_DRAWCLIPBOARD:
        begin
          ForwardMessage(AMsg);
          ClipboardChanged;
        end;
      WM_DESTROY:
        Enabled := False;
    else
      Result := DefWindowProc(FWindowHandle, Msg, WParam, LParam);
    end;
  end;
end;
```

**Comment by jasonjac2** [comment]
I was offering to send you the exe and then you could see the debug output in debugview (sysinternals) to see how it reacts to different clipboard events.

**Comment by jasonjac2** [comment]
although you are more than welcome to the code :-)

**Comment by stevencohn** [comment]
@jasonjac2 Is this still a problem for you? It seems to be an isolated case.

**Comment by jasonjac2** [comment]
Yes, unfortunately and only with OneMore.  
I have a todo to:
- delay my opening of the clipboard in case you / MS is opening it multiple times (you were clear you don't, but I am clutching at straws).

**Comment by stevencohn** [comment]
Fix in #1071
